x/ratelimit/README.md

```diff
@@ -0,0 +1,189 @@
+6. Finally, at the end of the 24 hours, the `Inflow` and `Outflow` would get reset to 0 and the `ChannelValue` would be re-calculated. In this example, the new channel value would be 104 (since more `uosmo` was sent to Stride, and thus more `ibc/uosmo` was minted)
+
+| Step |            Description           | Transfer Status | Inflow | Outflow | Net Inflow | Net Outflow | Channel Value |
+|:----:|:--------------------------------:|:---------------:|:------:|:-------:|:----------:|:-----------:|:-------------:|
+|   1  |        Rate limit created        |                 |    0   |    0    |            |             |      100      |
+|   2  |      8usomo Osmosis → Stride     |    Successful   |    8   |    0    |     8%     |             |      100      |
+|   3  |      8usomo Osmosis → Stride     |     Rejected    |   16   |    0    | 16% (>10%) |             |      100      |
+|   3  | State reverted after rejected Tx |                 |    8   |    0    |     8%     |             |      100      |
+|   4  |   12ibc/uosmo Stride → Osmosis   |    Successful   |    8   |    12   |            |      4%     |      100      |
+|   5  |      8usomo Osmosis → Stride     |    Successful   |   16   |    12   |     4%     |             |      100      |
+|   6  |            Quota Reset           |                 |    0   |    0    |            |             |      104      |
+
+
+## Denoms
+We always want to refer to the channel ID and denom as they appear on Stride. For instance, in the example above, we would store the rate limit with denom `ibc/D24B4564BCD51D3D02D9987D92571EAC5915676A9BD6D9B0C1D0254CB8A5EA34` and `channel-5`, instead of `uosmo` and `channel-326` (the ChannelID on Osmosis).
+
+However, since the ratelimit module acts as middleware to the transfer module, the respective denoms need to be interpreted using the denom trace associated with each packet. There are a few scenarios at play here...
+
+### Send Packets
+The denom that the rate limiter will use for a send packet depends on whether it was a native token (e.g. ustrd, stuatom, etc.) or non-native token (e.g. ibc/...)...
+#### Native vs Non-Native
+* We can identify if the token is native or not by parsing the denom trace from the packet
+    * If the token is **native**, it **will not** have a prefix (e.g. `ustrd`)
+    * If the token is **non-native**, it **will** have a prefix (e.g. `transfer/channel-X/uosmo`)
+#### Determining the denom in the rate limit
+* For **native** tokens, return as is (e.g. `ustrd`)
+* For **non-native** tokens, take the ibc hash (e.g. hash `transfer/channel-X/uosmo` into `ibc/...`)
+
+### Receive Packets
+The denom that the rate limiter will use for a receive packet depends on whether it was a source or sink
+#### Source vs Sink
+* **Source**: The packet is being received by a chain it was just sent from (i.e. the token has gone back and forth)
```

Comment:
I found the spec particularly confusing. That was my (failed) attempt to try and clarify it. 

How about this- kind of a hybrid of what I had and what you sent...

---

### Receive Packets
The denom that the rate limiter will use for a receive packet depends on whether it was a source or sink.

#### Source vs Sink
As a token travels across IBC chains, its path is recorded in the denom trace. 

* **Sink**: If the token moves **forward**, to a chain different than its previous hop, the destination chain acts as a **sink zone**, and the new port and channel are **appended** to the denom trace.
    * Ex1: `uatom` is sent from Cosmoshub to Stride 
      * Stride is the first destination for `uatom`, and acts as a sink zone
      * The IBC denom becomes the hash of: `/{stride-port)/{stride-channel}/uatom`
    * Ex2: `uatom` is sent from Cosmoshub to Osmosis then to Stride
      * Here the receiving chain (Stride) is not the same as the previous hop (Cosmoshub), so Stride, once again, is acting as a sink zone
      *  The IBC denom becomes the hash of: `/{stride-port)/{stride-channel}/{osmosis-port}/{osmosis-channel}/uatom`
   
* **Source**: If the token moves **backwards** (i.e. revisits the last chain it was sent from), the destination chain is acting as a **source zone**, and the port and channel are **removed** from the denom trace - undoing the last hop. Should a token reverse its course completely and head back along the same path to its native chain, the denom trace will unwind and reduce back down to the original base denom.
    * Ex1: `ustrd` is sent from Stride to Osmosis, and then back to Stride 
      * Here the trace reduces from `/{osmosis-port}/{osmosis-channel}/ustrd` to simply `ustrd`
    * Ex2: `ujuno` is sent to Stride, then to Osmosis, then back to Stride 
      * Here the trace reduces from `/{osmosis-port}/{osmosis-channel}/{stride-port}/{stride-channel}/ujuno` to just `/{stride-port}/{stride-channel}/ujuno` (the Osmosis hop is removed)
    * Stride is the source in the examples above because the token went back and forth from Stride -> Osmosis -> Stride

For a more detailed explanation, see the[ ICS-20 ADR](https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-001-coin-source-tracing.md#example) and [spec](https://github.com/cosmos/ibc/tree/main/spec/app/ics-020-fungible-token-transfer).